File: examples/picoenvmon/java/picoenvmon/data/LatestReadings.java
// SPDX-License-Identifier: GPL-3.0-only
package picoenvmon.data;

import picodroid.hardware.Sensor;

/**
 * Latest 1 Hz smoothed value per sensor type — written by {@code SensorLoggerService}'s smoothing
 * emit, read by the HTTP dashboard (network thread) and the UI. Plain unsynchronized fields: float
 * slots are written whole, and FreeRTOS scheduling on this single-core target makes the
 * cross-thread reads benign (a reader sees either the previous or the current sample, never a torn
 * one — floats occupy one 32-bit slot).
 */
public class LatestReadings {
  public static final int IDX_TEMPERATURE = 0;
  public static final int IDX_HUMIDITY = 1;
  public static final int IDX_PRESSURE = 2;
  public static final int IDX_GAS = 3;
  public static final int IDX_LIGHT = 4;
  public static final int COUNT = 5;

  private final float[] values = new float[COUNT];
  private int validMask;

  /** Map a {@link Sensor} type to an index, or -1. */
  public static int indexForType(int sensorType) {
    switch (sensorType) {
      case Sensor.TYPE_AMBIENT_TEMPERATURE:
        return IDX_TEMPERATURE;
      case Sensor.TYPE_RELATIVE_HUMIDITY:
        return IDX_HUMIDITY;
      case Sensor.TYPE_PRESSURE:
        return IDX_PRESSURE;
      case Sensor.TYPE_GAS_RESISTANCE:
        return IDX_GAS;
      case Sensor.TYPE_LIGHT:
        return IDX_LIGHT;
      default:
        return -1;
    }
  }

  public void updateByType(int sensorType, float value) {
    int idx = indexForType(sensorType);
    if (idx >= 0) {
      values[idx] = value;
      validMask |= (1 << idx);
    }
  }

  /** Whether {@code idx} has received at least one sample. */
  public boolean isValid(int idx) {
    return (validMask & (1 << idx)) != 0;
  }

  public float get(int idx) {
    return values[idx];
  }
}

File: examples/picoenvmon_kt/kotlin/picoenvmonkt/data/LatestReadings.kt
// SPDX-License-Identifier: GPL-3.0-only
package picoenvmonkt.data

import javax.inject.Inject
import javax.inject.Singleton
import picodroid.hardware.Sensor
import picoenvmonkt.IDX_GAS
import picoenvmonkt.IDX_HUMIDITY
import picoenvmonkt.IDX_LIGHT
import picoenvmonkt.IDX_PRESSURE
import picoenvmonkt.IDX_TEMPERATURE
import picoenvmonkt.READING_COUNT

/**
 * Latest 1 Hz smoothed value per sensor type — written by `SensorLoggerService`'s smoothing emit,
 * read by the HTTP dashboard (network thread) and the UI. Plain unsynchronized fields: float slots
 * are written whole, and FreeRTOS scheduling on this single-core target makes the cross-thread
 * reads benign (a reader sees either the previous or the current sample, never a torn one — floats
 * occupy one 32-bit slot).
 */
@Singleton
class LatestReadings @Inject constructor() {
    private val values = FloatArray(READING_COUNT)
    private var validMask = 0

    /** Map a [Sensor] type to an index, or -1. */
    private fun indexForType(sensorType: Int): Int =
        when (sensorType) {
            Sensor.TYPE_AMBIENT_TEMPERATURE -> IDX_TEMPERATURE
            Sensor.TYPE_RELATIVE_HUMIDITY -> IDX_HUMIDITY
            Sensor.TYPE_PRESSURE -> IDX_PRESSURE
            Sensor.TYPE_GAS_RESISTANCE -> IDX_GAS
            Sensor.TYPE_LIGHT -> IDX_LIGHT
            else -> -1
        }

    fun updateByType(sensorType: Int, value: Float) {
        val idx = indexForType(sensorType)
        if (idx >= 0) {
            values[idx] = value
            validMask = validMask or (1 shl idx)
        }
    }

    /** Whether `idx` has received at least one sample. */
    fun isValid(idx: Int): Boolean = (validMask and (1 shl idx)) != 0

    fun get(idx: Int): Float = values[idx]
}
